src/partition.c

```diff
@@ -1349,13 +1359,348 @@ static void pick_metis(int nodeID, struct space *s, int nregions,
+#endif
+    }
+    /* Define the cell graph. Keeping the edge weights association. */
+    int nadjcny = 0;
+    int nxadj = 0;
+    graph_init_scotch(s, s->periodic, weights_e, adjncy, &nadjcny, xadj,
+                      &nxadj);
+    /* Define the cell graph. Keeping the edge weights association. */
+    // Setting up the Scotch graph
+    SCOTCH_Graph graph;
+    SCOTCH_Num baseval = 0;
+    SCOTCH_Num vertnbr = ncells;         /* Number of vertices */
+    SCOTCH_Num edgenbr = (26 * vertnbr); /* Number of edges (arcs)   */
+
+    SCOTCH_Num *verttab; /* Vertex array [vertnbr + 1] */
+    verttab = (SCOTCH_Num *)malloc((vertnbr + 1) * sizeof(SCOTCH_Num));
+
+    // SCOTCH_Num *vendtab;   /* Vertex array [vertnbr] */
+    // vendtab = (SCOTCH_Num*) malloc((vertnbr) * sizeof(SCOTCH_Num));
+
+    SCOTCH_Num *velotab; /* Vertex load array        */
+    velotab = (SCOTCH_Num *)malloc((vertnbr) * sizeof(SCOTCH_Num));
+
+    SCOTCH_Num *edgetab; /* Edge array [edgenbr]     */
+    edgetab = (SCOTCH_Num *)malloc((edgenbr) * sizeof(SCOTCH_Num));
+
+    SCOTCH_Num *edlotab; /* Int load of each edge     */
+    edlotab = (SCOTCH_Num *)malloc((edgenbr) * sizeof(SCOTCH_Num));
+
+    int edges_deg = 26;
+    for (int i = 0; i < vertnbr; i++) {
+      verttab[i] = i * edges_deg;
+      velotab[i] = weights_v[i];
+    }
+    verttab[vertnbr] = vertnbr * edges_deg;
+
+    int vertex_count = 0;
+    int neighbour;
+    int return_edge;
+    /* The bidirectional weights associated with an edge are summed to ensure
+       that the resultant edges are symmetric. This is a neccessary for a Scotch
+       graph. */
+    for (int i = 0; i < edgenbr; i++) {
+      if ((i > (edges_deg - 1)) && (i % edges_deg == 0)) {
+        vertex_count++;
+      }
+      neighbour = adjncy[i];
+      edgetab[i] = neighbour;
+      for (int j = 0; j < edges_deg; j++) {
+        if ((adjncy[(neighbour * edges_deg + j)]) == vertex_count) {
+          return_edge = (neighbour * edges_deg + j);
+        }
+      }
+      edlotab[i] = weights_e[i] + weights_e[return_edge];
+    }
+
+    SCOTCH_graphInit(&graph);
+
+    if (SCOTCH_graphBuild(&graph, baseval, vertnbr, verttab, NULL, velotab,
+                          NULL, edgenbr, edgetab, edlotab) != 0) {
+      error("Error: Cannot build Scotch Graph.\n");
+    }
+    // #ifdef SWIFT_DEBUG_CHECKS
+    SCOTCH_graphCheck(&graph);
+    static int partition_count = 0;
+    char fname[200];
+    sprintf(fname, "scotch_input_com_graph_%03d.grf", partition_count++);
+    FILE *graph_file = fopen(fname, "w");
+    if (graph_file == NULL) {
+      printf("Error: Cannot open output file.\n");
+    }
+
+    if (SCOTCH_graphSave(&graph, graph_file) != 0) {
+      printf("Error: Cannot save Scotch Graph.\n");
+    }
+    fclose(graph_file);
+    // #endif
+    /* Read in architecture graph. */
+    SCOTCH_Arch archdat;
+    /* Load the architecture graph in .tgt format */
```

Comment:
Structurally this part looks odd to me. I think we should really be opening this file once during the 
initialisation and then caching whatever is needed for the constructions here.